File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        #find middle
        slow,fast = head,head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        prev = None
        #reverse from middle
        while slow:
            slow_nxt = slow.next
            slow.next = prev
            prev = slow
            slow = slow_nxt
        max_twn_sum=0
        #find sum
        while head and prev:
            max_twn_sum = max(max_twn_sum,head.val+prev.val)
            head = head.next
            prev = prev.next
        return max_twn_sum

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode kaju = null;
        while(slow.next != null){
            ListNode next = slow.next.next;
            slow.next.next = kaju;
            kaju = slow.next;
            slow.next = next;
        }  
        int curSum = 0;
        int maxSum = 0;
        while(kaju != null){
            curSum = kaju.val + head.val;
            if(curSum>maxSum){
                maxSum = curSum;
            }
            kaju = kaju.next;
            head = head.next;
        }
    return maxSum;
    }
}








// By reversing whole linked list

    //   ListNode headTwo = head;
//       ListNode kaju = null;
//       while(headTwo != null){
//         ListNode next = headTwo.next;
//         headTwo.next = kaju;
//         kaju = headTwo;
//         headTwo = next;
//       }  
//       int curSum = 0;
//       int maxSum = 0;
//         while(head != null){
//             curSum = head.val + kaju.val;
//             if(curSum>maxSum){
//                 maxSum = curSum;
//             }
//             head = head.next;
//             kaju = kaju.next;
//         }
//    return maxSum;